pkg/p2p/peer/peer.go

```diff
@@ -316,17 +273,19 @@ func (p *Reader) readLoop() {
 			return
 		}
 
-		// TODO: error here should be checked in order to decrease reputation
-		// or blacklist spammers
-		startTime := time.Now().UnixNano()
-		if err := p.router.Collect(message); err != nil {
-			l.WithError(err).Error("message routing")
-		}
+		go func() {
+			// TODO: error here should be checked in order to decrease reputation
+			// or blacklist spammers
+			startTime := time.Now().UnixNano()
```

Comment:
It's not boilerplate to test such scenario so I raise it as a question before ensuring its a real thing.

Reply:
While it is true that messages will be handled in a random order due to this approach, at first thought, I don't think this will cause any issues. `Reader`s do not expect messages to come in order, and they shouldn't either. Once a message is received, its processing is handled in a separate goroutine to all others, which means that if both a `Ping` and `GetData` are sent at the same time, they will always end up in the right place.

Reply:
It may however be good to test this, which I will spend some effort on.

Reply:
I'd advice to test this with enough attention. Go scheduler is non-deterministic.

Reply:
Right, however it is also true that the `MessageProcessor` will never allow messages to be crossed up, as explained in the chat thread. As mentioned, it may be an issue for messages that are only of the same topic, but this is more component-specific and thus I think the peer layer shouldn't be concerned with ordering messages. An example of that is the `Chain`, which solved it with the `sequencer` component.

Reply:
This problematic scenario is expected to be reproducible with `GetCandidate` request-response messaging.

Reply:
Could you elaborate?

Reply:
Once you restore the part for `GetCandidate from the network`, please provide a test ensuring that GetCandidate request is not responded with non-candidate payload due to the concern stated above.

Reply:
Like I said, messages that are not intended for the recipient will not end up there. When a candidate is requested, the broker will listen on `topics.Candidate` for incoming messages, so there is no conceivable way a `Ping` would end up in the broker, for instance. Since the `peer` is also not aware of ongoing requests and responses, there should be no way of mixing up messages. I can provide the test if you prefer, but I don't see how this would be possible.

Reply:
If you want to say that it is safe to correlate request and response only on the topic type, this is pretty wrong assumption. I suggested on the stand-up to introduce request_id (like stream identifier in HTTP/2). if you want to support concurrency in P2P message handling that's the idiomatic approach. (just not sure if this is the idea of this PR)

Reply:
While I also do not see an immediate correlation problem, I also do agree that adding an ID to correlate requests and response is pretty inexpensive, would help solving potential problems and as an approach could be expanded in the future. 

However, this is not part of this PR. @ggoranov please open an issue for this and detail how an idiomatic approach would look like

Reply:
I'm not just assuming it though; this is testable through the keepalive test harness, where request-response communication happens when transactions are created and blocks are produced. The handling of these kinds of messages, on a high level, has not really changed with this refactor, so I don't know why this wasn't brought up earlier. Where a `topics.Block` was originally passed through a channel by the `Synchronizer`, it is now handled through a callback in a goroutine. In both scenarios, the block processing is happening concurrently to the peer reading out new messages. 

If the peer reads out a message with `topics.GetCandidate`, it is processed by a callback which is registered to that topic, not the callback for `topics.Block` or anything else. The response topic is entirely dictated by the callback that is registered, so it would be a programming error if the callback responded on the wrong topic, and not an ephemeral one. If the message does not match the topic, the unmarshalling of the message should fail, and thus not even be processed further. So, I'm not just saying it is safe on a wild assumption.

However, as I seem unable to convince you of its safety, I will add a test once the candidate request procedure is added back.

Reply:
I also trust no-one without a test ;)